src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -185,6 +194,62 @@ public void enableCollection(final MetricType metricType, final boolean enable)
         this.topQueriesServices.get(metricType).setEnabled(enable);
     }
 
+    /**
+     * Validate grouping given grouping type setting
+     * @param groupingTypeSetting grouping setting
+     */
+    public void validateGrouping(final String groupingTypeSetting) {
+        GroupingType.getGroupingTypeFromSettingAndValidate(groupingTypeSetting);
+    }
+
+    /**
+     * Set grouping
+     * @param groupingTypeSetting grouping
+     */
+    public void setGrouping(final String groupingTypeSetting) {
+        GroupingType newGroupingType = GroupingType.getGroupingTypeFromSettingAndValidate(groupingTypeSetting);
+        GroupingType oldGroupingType = groupingType;
+
+        if (oldGroupingType != newGroupingType) {
+            groupingType = newGroupingType;
+
+            for (MetricType metricType : MetricType.allMetricTypes()) {
+                this.topQueriesServices.get(metricType).setGrouping(newGroupingType);
+            }
+        }
+    }
+
+    /**
+     * Set max number of groups
+     * @param maxGroups maximum number of groups that should be tracked when calculating Top N groups
+     */
+    public void setMaximumGroups(final int maxGroups) {
+        for (MetricType metricType : MetricType.allMetricTypes()) {
+            this.topQueriesServices.get(metricType).setMaxGroups(maxGroups);
+        }
+    }
+
+    /**
+     * Validate max number of groups. Should be between 1 and MAX_GROUPS_LIMIT
+     * @param maxGroups maximum number of groups that should be tracked when calculating Top N groups
+     */
+    public void validateMaximumGroups(final int maxGroups) {
+        if (maxGroups < 1 || maxGroups > QueryInsightsSettings.MAX_GROUPS_LIMIT) {
```

Comment:
If `maxGroups` < N it can lead to unexpected results. Should we validate that here?

Reply:
maxGroups only kicks in after the minimum queue is filled. So first we fill in the min queue for Top N. We then start overflowing groups to the max queue. The maxGroups will limit the size of the max queue.